provisioning/roles/base/tasks/perl_configuration.yml

```diff
@@ -22,3 +13,15 @@
   become: yes
   with_items:
     - { key: "defaultRSVPServer", value: "{{ rsvp_server }}" }
+
+- name: Configure Permabit::RSVP multi-level list values in perl.yaml
+  replace:
+    path: /etc/permabit/perl.yaml
+    regexp: "{{ item.keyRegex }}"
+    replace: "{{ item.replacement }}"
+  become: yes
+  with_items:
+      # (?s) -> . matches newline
+      # (\s+) -> capture whitespace before key to indent new values correctly
+    - { keyRegex: '(?s)(^Permabit::RSVP:.+?config:.+?processes:.+?ok:.+?\n(\s+)user:\n)',
+        replacement: '\1\2  - ssh .* sftp\n\2  - sshfs\n' }
```

Comment:
> So this is adding entries under `user:` in the config, right?
> 
> ```
>   - ssh .* sftp
>   - sshfs
> ```
> 

Yes.

> This is a practice that I think will get really hard to maintain over time just due to the complexity of the regex. I think if we need to do much more than this in the future, then we should consider replacing the text with something more templated and easy on the eyes.
> 
> In the previous spelling, we looked specifically for `#` and made sure to not put it back in. I'm assuming that's what the `.+?` is doing, but I'm curious if there are other things it might also be matching and either dropping or maintaining incorrectly?

That's the reason the regex is what it is.  It establishes the context from the class being configured to the appropriate entry w/o greedily consuming (`.+?`) the contents of the file.  It's the equivalent of (in the case of the yaml having been loaded into memory):`yaml["Permabit::RSVP"]["config"]["processes"]["ok"]["user"]`.

While there are ways to load/write a yaml file from within ansible they do not preserve comments and consequently are problematic for other tasks that may need to manipulate the file.

Reply:
Improved idiot proofing.

Reply:
I resemble that comment. ;)